File: csharp/word-search/WordSearch.cs
using System;
using System.Collections.Generic;
using System.Linq;

public class WordSearch
{

    private Dictionary<char, List<(int, int)>> gridDict;
    private readonly (int, int)[] directions = new[] { (1,0),(1,1),(0,1),(-1,1),(-1,0),(-1,-1),(0,-1),(1,-1) };
    public WordSearch(string grid)
    {
        gridDict = new Dictionary<char, List<(int, int)>>();
        string[] gridRows = grid.Split('\n', StringSplitOptions.TrimEntries).ToArray();
        for (int j = 0; j < gridRows.Length; j++)
        {
            for (int i = 0; i < gridRows[j].Length; i++)
            {
                (int, int) actualPos = (i + 1, j + 1);
                char actualChar = gridRows[j][i];
                if (gridDict.ContainsKey(actualChar))
                {
                    gridDict[actualChar].Add(actualPos);
                } 
                else
                {
                    gridDict.Add(actualChar, new List<(int, int)> { actualPos });
                }
            }
        }
    }

    public Dictionary<string, ((int, int), (int, int))?> Search(string[] wordsToSearchFor) =>
        wordsToSearchFor.Select(word => CheckWord(word)).ToDictionary(el => el.Item1, el => el.Item2);

    private (string, ((int, int),(int, int))?) CheckWord(string word)
    {
        if (word.Any(c => !gridDict.ContainsKey(c))) { return (word, null); }
        
        IEnumerable<((int, int), (int, int))> results = gridDict[word[0]]
            .SelectMany(start => directions.Select(dir => (start, dir)))
            .Where(comb => CheckDirection(word, comb.start, comb.dir));
        
        if (results.Any())
        {
            (int, int) start = results.First().Item1;
            (int, int) dir = results.First().Item2;
            (int, int) endPos = addTuples(start, scalarMult(word.Length - 1, dir));
            return (word, (start, endPos));
        }
        return (word, null);
    }

    private bool CheckDirection(string word, (int, int) start, (int, int) dir)
    {
        for (int i = 1; i < word.Length; i++)
        {
            (int, int) pos = addTuples(start, scalarMult(i, dir));
            if (!gridDict[word[i]].Contains(pos)) { return false; }
        }
        return true;
    }

    private (int, int) addTuples((int, int) t1, (int, int) t2) => (t1.Item1 + t2.Item1, t1.Item2 + t2.Item2);
    private (int, int) scalarMult(int k, (int, int) t) => (k * t.Item1, k * t.Item2);
}

File: fsharp/word-search/WordSearch.fs
module WordSearch

let directions = seq { (1,0);(1,1);(0,1);(-1,1);(-1,0);(-1,-1);(0,-1);(1,-1) }

// helper functions for addition and scalar multiplication of tuples
let addTuples (a1,b1) (a2,b2) = (a1+a2,b1+b2)
let scalarMult k (a,b) = (k*a,k*b)

// no Map.merge in the F# Map module, what a shame...
// Imitating the pattern of Map.merge in Elixir,
// we take the first map a as a starting point and fold over every entry (k, v) of the second map b
// while resolving merge conflicts by application of a function f with parameters k, value_in_a and value_in_b
let merge (a : Map<'a, 'b>) (b : Map<'a, 'b>) (f : 'a -> 'b -> 'b -> 'b) =
    (a, b) ||> Map.fold (fun acc k vb ->
        match Map.tryFind k acc with
        | Some va -> acc |> Map.add k (f k va vb)
        | None -> acc |> Map.add k vb)

// collects for every given character in the grid (the key of the map) a sequence of its index positions (the value of the map) 
let gridToMap (grid : string list) : Map<char,(int * int) seq> =
    grid 
    |> Seq.indexed 
    |> Seq.map (fun (i, s) -> 
        s 
        |> Seq.indexed 
        |> Seq.map (fun (j, c) -> 
            seq { (c, seq { (j+1, i+1) }) } |> Map.ofSeq 
        )
        |> Seq.fold (fun acc m -> merge acc m (fun _ v1 v2 -> Seq.append v1 v2)) Map.empty
     )
     |> Seq.fold (fun acc m -> merge acc m (fun _ v1 v2 -> Seq.append v1 v2)) Map.empty

// fetches all possible starting positions from the gridMap by looking at the first character of the word
// then enriches every starting position with all possible 8 directions and then uses checkDir from index 1 
// to filter out the successful combinations, takes the first successful result (if any), 
// calculates the end position of the word and wraps start and end position together in an option
let checkWord (gridMap : Map<char,(int * int) seq>) (word : string) : string * Option<((int * int) * (int * int))> = 
    let rec checkDir start dir index = 
        if index = word.Length then true 
        else 
            let position = addTuples start (scalarMult index dir) 
            if gridMap[word[index]] |> Seq.contains position then checkDir start dir (index + 1) else false 

    if word |> Seq.exists(fun ch -> Map.containsKey ch gridMap |> not) then (word, None)
    else 
        let results = 
            gridMap[word[0]] 
            |> Seq.collect (fun start -> directions |> Seq.map (fun dir -> (start, dir)))
            |> Seq.filter (fun (start, dir) -> checkDir start dir 1)
        match results |> Seq.tryHead with 
            | None -> (word, None)
            | Some (start, dir) -> 
                let endpos = addTuples start (scalarMult (word.Length - 1) dir) 
                (word, Some (start, endpos))
      
let search grid wordsToSearchFor = 
    let gridMap = gridToMap grid 
    wordsToSearchFor |> Seq.map (fun word -> checkWord gridMap word) |> Map.ofSeq
